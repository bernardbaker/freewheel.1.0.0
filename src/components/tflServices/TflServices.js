import React, {Component} from 'react'
import _ from 'underscore'
import './TflServices.css'

class TflServices extends Component {

  constructor(props) {
    super(props)
    
    this.state = {
      data: []
    }
  }

  componentDidCatch(error, info) {
    console.log(error)
    console.log(info)
  }

  static getDerivedStateFromProps(nextProps, prevState){
    let values = _.sortBy(nextProps.data, mode => [mode.modeName, mode.name])
    if(values.length > 0) {
      values.splice(0, 0, {'info':'Please select a service'})
      values.push({'name':'Cycle Hire'})
      return {
        data : values
      }
    }
    return null
  }

  render() {
    return (
      <div>
        <h1>TFL Services</h1>
          <select
            ref="select"
            className="Tfl-Service-List"
            onChange={this.props.handleSelection}
            id="service-types"
            key="service-types"
            value={this.props.selectedIndex}
          >
          {
            this.state.data.map((element, index, array) => {
              if(index === 0) {
                return <option
                  key={'info-'+index}
                  >
                  {element.info}
                  </option>
              }
              else if(index < array.length-1) {
                let modeName = "".concat(
                  element.modeName.toString().charAt(0).toUpperCase(),
                  element.modeName.toString().substr(1),
                  ' - ',
                  element.name,
                  (element.serviceTypes.filter(service => service.name === 'Night').length > 0 ? ' - Runs at night' : '')
                )
                const styleDisruption = element.lineStatuses.filter(status => status.statusSeverity !== 10).length > 0 ? 'Red' : 'Green'
                return <option
                  key={'service-'+index}
                  className={styleDisruption}
                  data-category={element.id}
                  data-type="service-update"
                  value={element.name.toString().toLowerCase().split(' ').join('')}
                  >
                  {modeName}
                  </option>
              } else {
                return <option
                  key={'service-'+index}
                  data-category="cycle-hire"
                  data-type="cycle-hire"
                  value="cycle-hire"
                  >
                  {element.name}
                  </option>
              }
              
            })
          }
          </select>
      </div>
    )
  }
}

export default TflServices