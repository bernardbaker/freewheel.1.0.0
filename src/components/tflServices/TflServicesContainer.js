import React from 'react'

export default function (Component) {
  return class Container extends React.Component {

    constructor(props) {
      super(props)
      this.state = {
        data: [],
        selectedIndex: 0
      }

      this.handleSelection = this.handleSelection.bind(this)
    }

    handleSelection(event) {
      const category = event.target.options[event.target.selectedIndex].getAttribute('data-category')
      const type = event.target.options[event.target.selectedIndex].getAttribute('data-type')
      this.setState({
        selectedIndex: event.target.options[event.target.selectedIndex].value
      })
      this.props.serviceSelectedAction(category, type, this.state.data)
    }
    
    render() {
      return <Component {...this.props} data={this.state.data} handleSelection={this.handleSelection} selectedIndex={this.state.selectedIndex} />
    }

    componentDidCatch(error, info) {
      console.log(error)
      console.log(info)
    }

    componentDidMount() {
      fetch('https://api.tfl.gov.uk/Line/Mode/tube,overground,dlr/Status?detail=true')
      .then(response => {
        let res = response.json();
        res.then(payload => {
          this.setState(prevState => ({
              data: prevState.data = payload
          }))
        })
      })
    }
  }  
}