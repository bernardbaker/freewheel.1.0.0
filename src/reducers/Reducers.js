export default (state = {}, action) => {
  switch (action.type) {
    case 'SERVICE_SELECTED_ACTION':
      return Object.assign({}, state, {
        category: action.payload.category,
        service: action.payload.service,
        data: action.payload.data
      })
    default:
      return state
  }
}