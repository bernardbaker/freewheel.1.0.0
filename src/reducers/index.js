import { combineReducers } from 'redux';
import reducers from './Reducers';
export default combineReducers({
  reducers
});