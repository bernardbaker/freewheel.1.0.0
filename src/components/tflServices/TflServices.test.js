import React from 'react';
import ReactDOM from 'react-dom';
import TflServicesContainer from './TflServicesContainer';
import TflServices from './TflServices'
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer'

import { createStore } from 'redux'
import reducers from '../../reducers'  

const store = createStore(reducers)

const DELAY_MS = 2000

const sleep = (ms) => {
  return new Promise(resolve => setTimeout(resolve, ms))
}

describe('TflServicesContainer component', () => {
  it('matches the snapshot', () => {
    const Component = TflServicesContainer(TflServices)
    const tree = renderer.create(<Component />).toJSON()
    expect(tree).toMatchSnapshot()
  })
  
  it('default data when created', () => {
    const Component = TflServicesContainer(TflServices)
    const wrapper = shallow(<Component />)
    expect(wrapper.state().data).toEqual([])
  })
  
  it('call for data when mounted', async () => {
    const ORIGINAL_TIMEOUT = jasmine.DEFAULT_TIMEOUT_INTERVAL
    jasmine.DEFAULT_TIMEOUT_INTERVAL = 10000
    const Component = TflServicesContainer(TflServices)
    const wrapper = shallow(<Component />)
    await wrapper.instance().componentDidMount()
    await sleep(DELAY_MS + 5000)
    expect(wrapper.state().data.length).toBeGreaterThan(0)
    jasmine.DEFAULT_TIMEOUT_INTERVAL = ORIGINAL_TIMEOUT
  })
  
})