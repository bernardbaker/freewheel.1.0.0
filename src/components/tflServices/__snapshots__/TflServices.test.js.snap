// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TflServicesContainer component matches the snapshot 1`] = `
<div>
  <h1>
    TFL Services
  </h1>
  <select
    className="Tfl-Service-List"
    id="service-types"
    onChange={[Function]}
    value={0}
  />
</div>
`;
